compiler/codegen/mli.cpp

```diff
@@ -0,0 +1,912 @@
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    if (isSupportedType(as->type)) { continue; }
+    USR_FATAL("MLI does not support code generation for type", as->type);
+  }
+
+  return;
+}
+
+Type* MLIContext::getTypeFromFormal(ArgSymbol* as) {
+  if (as == NULL) { return NULL; }
+  return as->type;
+}
+
+Type* MLIContext::getTypeFromFormal(FnSymbol* fn, int i) {
+  return getTypeFromFormal(fn->getFormal(i));
+}
+
+std::string MLIContext::genClientsideRPC(FnSymbol* fn) {
+  bool hasVoidReturnType = fn->retType == dtVoid;
+  bool hasFormals = fn->numFormals() != 0;
+  std::string gen;
+
+  // Declare the unique ID for this function.
+  gen += "int64_t id = ";
+  gen += this->genFuncNumericID(fn);
+  gen += ";\n";
+
+  // Declare a int64 for the server return value.
+  gen += this->genNewDecl("int64_t", "st");
+
+  // Declare a temporary for the return value, if necessary.
+  if (not hasVoidReturnType) {
+    gen += this->genNewDecl(fn->retType, "result");
+  }
+
+  if (this->injectDebugPrintlines) {
+    gen += this->genDebugPrintCall(fn, "[Client]");
+  }
+
+  // Push function to call.
+  gen += this->genSocketPushCall(client_main, "id");
+
+  // Pull server confirmation.
+  gen += this->genSocketPullCall(client_main, "st");
+
+  // TODO: Handle server errors.
+  gen += this->genTodo("Handle server errors.");
+  gen += "if (st) { ;;; }\n";
+
+
+  // If we are void/void, then there's nothing left to do.
+  if (hasVoidReturnType and not hasFormals) {
+    gen += this->genComment("Routine is void/void!");
+    return gen;
+  }
+
+  // Issue pack call for each formal.
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    ArgSymbol* as = fn->getFormal(i);
+    Type* t = getTypeFromFormal(as);
+
+    gen += this->genMarshalPushCall(client_arg, as->name, t);
+  }
+
+  // Pull and return result if applicable.
+  if (not hasVoidReturnType) {
+    gen += "result = ";
+    gen += this->genMarshalPullCall(client_res, "result", fn->retType);
+    gen += "return result;\n";
+  }
+ 
+  return gen;
+}
+
+std::string MLIContext::genServersideRPC(FnSymbol* fn) {
+  std::map<int, std::string> formalTempNames;
+  bool hasVoidReturnType = fn->retType == dtVoid;
+  bool hasFormals = fn->numFormals() != 0;
+  std::string gen;
+
+  // Emit void/void calls immediately, then return.
+  if (hasVoidReturnType and not hasFormals) {
+    gen += fn->cname;
+    gen += "();\n";
+    return gen;
+  }
+
+  // Declare a temporary for the return value, if necessary.
+  if (not hasVoidReturnType) {
+    gen += this->genTypeName(fn->retType);
+    gen += " result;\n";
+  }
+
+  // Declare temporaries, issue unpack call for each formal.
+  for (int i = 1; i <= fn->numFormals(); i++) {
+    std::string tmp;
+
+    Type* t = this->getTypeFromFormal(fn, i);
+
+    gen += this->genTypeName(t);
+    gen += " ";
+
+    // Map temp names to formal indices (shifted down one).
+    tmp += "tmp_";
+    tmp += std::to_string(i - 1);
```

Comment:
Why not use the formal's name?  That seems like it would be easier to understand when reading the generated code

Reply:
It was just preference. If you'd rather me mangle it in that fashion, I can.

Reply:
I think I would prefer it, thanks!